Upgrade to Gradle 6.7.1

karivatj opened the pull request:
Hello,

I've been working on refactoring the android-gpgs implementation because it uses deprecated Google APIs. However, when building the project I noticed that the project is based on an older Gradle version. I figured I'd migrate the project to Gradle 7.0.3 before proceeding with the actual refactoring.

Here is a PR that contains the needed changes to build / deploy the project with Gradle 7. Basically all of the gradle scripts needed some tuning to get this to work.

Building with `gradlew clean publishToMavenLocal` on Windows or `gradle clean publishToMavenLocal` on Linux works great and the local repository populates with libraries I can use in my project. 

Obviously I haven't tested deploying the package to the external repository but thats something maybe you, the developer, can test.

If you need me to do any additional fixes or other changes please let me know.

MrStahlfelge commented:
Hi, 
please repair the GitHub actoins build.

There are some changes that seem unnecessary (renaming minSdk, targetSdk etc. to *Ver) and cause changes that could be avoided.

Setting ios and core modules to Java version 8 is dangerous until all platforms supported support Java 8. While RoboVM can handle some Java 8 language features, it is still better to keep the source level on 7 so problems get caught on compile time, not on runtime.

Are you sure Gradle 7 is needed? For most things, Gradle 6.7 is still a viable alternative which causes much less problems in terms of backwards compatibility.

karivatj commented:
Hi,

Thanks for the comments. I will take a look at the issues you raised and will update the PR accordingly.

Br.
Kari

karivatj commented:
Hello,

I looked into the issues you mentioned and here are some thoughts:

`There are some changes that seem unnecessary (renaming minSdk, targetSdk etc. to *Ver) and cause changes that could be avoided.`

I will go through the changes and remove any unnecessary changes. However, with the new Gradle 7 the `compileSdk`, `targetSdk `and `minSdk` are reserved keywords and thus cannot be used. That is why we need to use a different name for the variables.

`Setting ios and core modules to Java version 8 is dangerous until all platforms supported support Java 8. While RoboVM can handle some Java 8 language features, it is still better to keep the source level on 7 so problems get caught on compile time, not on runtime.`

Fair enough I will revert back to Java 7. I was refactoring the gpgs-client and decided to use lambdas which are supported in Java 8. I will revert the changes.

`Are you sure Gradle 7 is needed? For most things, Gradle 6.7 is still a viable alternative which causes much less problems in terms of backwards compatibility.`

I made some investigation and couldn't find any source that declares one shouldn't use Gradle 7. As far as I know Gradle 7 is the latest stable version and incorporates many improvements and bugfixes over Gradle 6. What is your preference? Should we stick to Gradle 7 or should I downgrade to Gradle 6?

karivatj commented:
Reverted to Java 7 and made a slight change to `compile.yml` 

Had to change:

`run: ./gradlew clean uploadArchives -PSNAPSHOT=true -PLOCAL=true`

to

`run: ./gradlew clean publishToMavenLocal -PSNAPSHOT=true -PLOCAL=true`

`publishToMavenLocal ` copies all defined publications to the local Maven cache, including their metadata (POM files, etc.).

When publishing a release version one should use

`run: ./gradlew clean publish -PRELEASE=true

MrStahlfelge commented:
Gradle 7 needs Java 11, so not useable on environments still running jdk 8. That's why 6.7 is widely used at the moment. I would prefer to use it if it works with the latest dependencies.

karivatj commented:
Hi,

Understood. I downgraded Gradle to 6.7.1 and `com.android.tools.build:gradle` module to version 4.2.2 which supports Gradle 6.7.1+.

Builds fine with `./gradlew clean publishToMavenLocal -PSNAPSHOT=true -PLOCAL=true`

MrStahlfelge commented on build.gradle:
```diff
@@ -1,17 +1,23 @@
+        maven { url "https://plugins.gradle.org/m2/" }
+        maven { url "https://oss.sonatype.org/content/repositories/snapshots/" }
+        maven { url 'https://developer.huawei.com/repo/' }
+        gradlePluginPortal()
```
I think this is the same as plugins.gradle.org/m2

karivatj replied:
Yeah I didn't realize this. I will remove line 11 as gradlePluginPortal is already defined.

MrStahlfelge commented on build.gradle:
```diff
@@ -1,17 +1,23 @@
         classpath 'org.wisepersist:gwt-gradle-plugin:1.0.6'
-        classpath 'com.android.tools.build:gradle:3.2.0'
+        classpath 'com.android.tools.build:gradle:4.2.2'
+        classpath 'com.google.gms:google-services:4.3.10'
```
I think it is better to leave dependencies that are only used for single modules in these modules. It is better documented where it is used this way.

karivatj replied:
I will look into this and move the dependencies inside the modules they are used.

karivatj replied:
I have moved the google services dependencies to android-gpgs submodule and huawei related stuff to android-huawei submodule. Should be ok now.

MrStahlfelge commented on build.gradle:
```diff
@@ -21,23 +27,34 @@ allprojects {
 
-        appName = 'gdx-gamesvcs'
-        gdxVersion = '1.9.8'
+        gdxVersion = '1.10.0'
```
Is there a problem with 1.9.8 that needs to raise to 1.10.0?

karivatj replied:
No problems with 1.9.8. I just figured that might as well raise the libGdx version to the newest stable version now that it is available. Is it okay to leave as is?

MrStahlfelge replied:
For the library, it is better to use the oldest version possible here due to how Gradle works when the library is used within game projects: Gradle will grab the newest version of all declared - so, in order to not change game projects without a need, let's keep 1.9.8.

karivatj replied:
I have reverted the version update back to 1.9.8

MrStahlfelge commented on .github/workflows/compile.yml:
```diff
@@ -25,4 +25,4 @@ jobs:
       uses: android-actions/setup-android@v2.0.2
     - name: Local install
-      run: ./gradlew clean uploadArchives -PSNAPSHOT=true -PLOCAL=true
+      run: ./gradlew clean publishToMavenLocal -PSNAPSHOT=true -PLOCAL=true
```
-PLOCAL=true is not needed according to your README change?

Also, what commands are needed for publishing a snapshot and a release with these changes?

karivatj replied:
Here is the reference docs for maven publishin: https://docs.gradle.org/current/userguide/publishing_maven.html

TLDR;

`publish`
An aggregate task that publishes all defined publications to all defined repositories. It does not include copying publications to the local Maven cache.

`publishToMavenLocal`
Copies all defined publications to the local Maven cache, including their metadata (POM files, etc.).

So basically you should be able to do` ./gradlew publish -PRELEASE=true -PSNAPSHOT=true` and it should attempt to publish the artifacts to both release and snapshot repos. 

`-PLOCAL` is not needed because `publishToMavenLocal` handles the publish procedure to a local repository

MrStahlfelge reviewed (commented):
Thanks for your work so far, please check my questions/comments.

karivatj commented:
Updated the PR. 

- Libgdx back to 1.9.8. 
- Moved dependencies to submodules where they are used. 
- Added some comments to the question about publishing the artifacts.

MrStahlfelge reviewed (approved):
This works well. There are some help methods in the publish.gradle files that could be removed now, but we can do that on another step.

Thanks for your efforts.
